lib/util/Utf8.v3

```diff
@@ -46,47 +48,55 @@ component Utf8 {
 		if (codepoint >= 0xD800 && codepoint <= 0xDFFF) return false; // UTF-16 surrogates
 		return true;
 	}
+	// Decode a UTF-8 sequences of bytes to one Unicode codepoint. 
+	// Returns (byte offset of the char, codepoint).
+	def decode1(str: Array<byte>, start: int, end: int) -> (int, u32) {
```

Comment:
Accepting a `Range<byte>` as input will make this a little more general and robust.

Also, what about returning `(-1, 0)` in case of error, instead of an invalid codepoint? That would be similar to how `Ints.parseDecimal()` works.

Reply:
> Also, what about returning (-1, 0) in case of error, instead of an invalid codepoint? That would be similar to how Ints.parseDecimal() works.

We need to return the byte offset of the invalid character, so there are several error cases. Only `-1` is not enough. What about changing the spec of the function as below?
```
def decode1(str: Array<byte>, r: Range<byte>, f: u32 -> void) -> (int, bool)
```
We return a bool value to indicate success or failure.

Reply:
@titzer Splited the tests. Should I change this too?

Reply:
Sorry for the late reply. If this function instead accepts a `Range<byte>`, I think its OK for this function to just return at `(u32, bool)` pair, because the caller can either assume byte 0 is wrong, or if finer granularity is needed, figure out which byte is wrong.